plymouth: attempt to make flicker-free

eadwu opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->

If you rig up some plymouth prompts it should work fine for decryption and messages now. Patches are a big mess.

Not sure what's actually needed in stage-1 for this to run as early as possible.
###### Motivation for this change
label-ft: https://gitlab.freedesktop.org/plymouth/plymouth/-/merge_requests/91

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

matthewbauer commented:
Awesome work! Haven't tried it yet, but this is definitely nice to have.

On stage 1, I think we may need to use the systemd integration more for this to work with plymouth. I'm not sure what this takes, but it had been brought up at one point.

Related to https://github.com/NixOS/nixpkgs/issues/26722 and https://github.com/NixOS/nixpkgs/issues/44965

eadwu commented:
Right now I'm using a makeshift wrapper for reading and showing messages, but the implementation is works only on bash due to the requirement on `type`.

```
    if [ ! "$(type -t _read)" = "function" ]; then
      _read() {
        local pwd
        pwd="$(plymouth ask-for-password "$@")"
        printf "$pwd"
      }
    fi

    if [ ! "$(type -t _prompt)" = "function" ]; then
      _prompt() {
        plymouth display-message "$@"
      }
    fi
```

but this ends up becoming far over-complicated because of the differing usages of outputting within initrd (or unless we enforce a strict format for the function, like for `_read` plymouth can only do `--text "$1"` and `_prompt` can only do `--text "$1"` which would probably work just fine).

Not sure on how to replicate non-newline displaying of messages, but it can be partially mitigated by having a sort of message history in the plymouth theme itself (bottom right on the image).
![Screenshot_2020-04-03_12-05-58](https://user-images.githubusercontent.com/22758444/78381646-aa917580-75a3-11ea-94ef-34d9fa59ace7.png)

worldofpeace commented:
cc @mkg20001

matthewbauer commented:
I'm getting this when building an OVA image:

```
output '/nix/store/d8yv4bc0kw9j3dg9fc3n2lp8qlljf6i4-extra-utils' is not allowed to refer to the following paths:
  /nix/store/il5jhmg7m01sflpz1g582dvwsii6byb9-plymouth-2020-03-25
```

eadwu commented:
Should've fixed. Seemed to be a problem that occurred with the default settings.

Tried booting it up with the default theme and it should work fine as long as the password issue is resolved for passwords (without bash-specific workarounds). I'll work on that when I'm able to work on the PR for an extended period of time.

The default theme also has a problem with the password prompt "bullets" which I think is probably due to the fact the fact doesn't contain the character.
```
local.bullet_image = WriteText("•", palette.text.contrast);
```

worldofpeace commented:
> Not sure what's actually needed in stage-1 for this to run as early as possible.

I believe from the people I've talked to about this, it would be https://github.com/NixOS/nixpkgs/pull/74842/.

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,61 +1,81 @@
 
   buildInputs = [
-    gtk3 udev systemd
+    gtk3 udev systemd freetype intltool
```
`intltool` should likely go to `nativeBuildInputs`.

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,61 +1,81 @@
     "--without-rhgb-compat-link"
+    "--without-system-root-install"
+
+    "--enable-drm"
```
Could you split clean ups and functional changes into a separate commit? It is hard to review otherwise.

eadwu replied:
I think I split it up so that it's at least somewhat more readable? But at least the functionality improvements should be distinct now.

eadwu committed:
plymouth: cleanup derivation

eadwu committed:
plymouth: 0.9.4 -> 2020-04-07

eadwu committed:
plymouth: include label-ft plugin

A lightweight alternative to the label plugin.

/usr/bin/fc-match is patched so that it'll be more NixOS-y compatible
but it is not expected that it will ever be used since after initrd,
label.so will be present and thus will be used instead.

https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/45
https://gitlab.freedesktop.org/plymouth/plymouth/-/merge_requests/91

eadwu committed:
nixos/modules/plymouth: cleanup

eadwu committed:
nixos/modules/plymouth: use upstream defaults

https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/101
https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/75204a2517658c01bd1d39f2dff845512c9cab12/src/plymouthd.defaults

eadwu committed:
nixos/modules/plymouth: add font option

In order to utilize the label-ft plugin, we need a font

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,61 +1,94 @@
+
+  patches = [
+    # https://build.opensuse.org/package/show/openSUSE%3AFactory/plymouth
+    ./manpages.patch
```
Why is this patch needed? Reading https://www.gnu.org/software/automake/manual/html_node/Man-Pages.html, it sounds like it should work as is.

eadwu replied:
Seems like it isn't, removed.

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,61 +1,94 @@
+    # ./only_use_fb_for_cirrus_bochs.patch
+  ] ++ stdenv.lib.optionals withLabelFt [
+    # https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/45
+    ./v3-0001-ply-label-Don-t-crash-if-label-plugin-fails.patch
```
Is this patch set necessary for flicker-free experience? I would rather not include unmerged patches using legacy libraries (freetype).

eadwu replied:
It isn't necessary but without it, some of the functionality for some themes will be impacted, anything that uses `display-message` won't be shown since we don't bundle the original `label.so` since it's far more bloated than `label-ft.so`.

But if it's just the password prompt (basically a pile of images) and the background, then no it isn't needed.

eadwu commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,61 +1,94 @@
-      -e "s#\$(\$PKG_CONFIG --variable=systemdsystemunitdir systemd)#$out/etc/systemd/system#g" \
-      -e "s#plymouthplugindir=.*#plymouthplugindir=/etc/plymouth/plugins/#" \
       -e "s#plymouththemedir=.*#plymouththemedir=/etc/plymouth/themes#" \
+      -e "s#plymouthplugindir=.*#plymouthplugindir=/etc/plymouth/plugins/#" \
```
Reverted.

eadwu commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,61 +1,94 @@
+  configurePlatforms = [ "host" ];
+
   configureFlags = [
-    "--sysconfdir=/etc"
```
I don't see any problems with the output so added.

berbiche commented:
How can I test this in my configuration?

davidak commented:
@berbiche you have also to enable plymouth, but then it should work.

berbiche commented on nixos/modules/system/boot/plymouth.nix:
```diff
@@ -16,14 +16,15 @@ let
+  configFile = writeText "plymouthd.conf" ''
     [Daemon]
     ShowDelay=0
+    DeviceTimeout=8
```
Should there be a way to make this configurable?

eadwu committed:
nixos: misc plymouth support

davidak commented:
I don't see a difference. https://youtu.be/wHbl62EgrlE

The stage1 text is still shown, then a black screen, and after 20 seconds i see the lightdm login.

i'm using Pantheon desktop
EFI systemd-boot
no disk encryption

should i test on a system with disk encryption? or still too early to test?

eadwu commented:
Took a quick glance at the related PRs and still doesn't look like it's possible. The earliest I went was stage-2 I believe (I don't remember)?
